hawc/apps/common/api/mixins.py

```diff
@@ -108,3 +110,113 @@ class LegacyAssessmentAdapterMixin:
+        assert self.read_serializer_class is not None, (
+            "'%s' should either include a `read_serializer_class` attribute,"
+            "or override the `get_read_serializer_class()` method." % self.__class__.__name__
+        )
+        return self.read_serializer_class
+
+    def get_write_serializer_class(self):
+        assert self.write_serializer_class is not None, (
+            "'%s' should either include a `write_serializer_class` attribute,"
+            "or override the `get_write_serializer_class()` method." % self.__class__.__name__
+        )
+        return self.write_serializer_class
+
+
+class GetOrCreateMixin:
+    """
+    this should be mixed into a serializer
+
+   This mixin:
+    1. disables any UniqueTogetherValidators
+    2. overrides create to actually call get_or_create
+    3. provides default validate behavior that can either be used as is
+       or called in conjunction with custom validation on the serializer mixing
+       it in (or bypassed completely if needed).
+
+    The serializer mixing this in should pass in a "context" argument containing
+    the requesting user, e.g.:
+
+    serializer = self.get_serializer( data=x, context={"user": request.user} )
+
+    that user will be used by the default validation behavior, and stored as 
+    self.requesting_user if any custom validation needs to be done based on the
+    caller.
+
+    The end result is a serializer mixing this in can either lookup or create
+    an element (or elements, works fine for multiples), so 
+    that clients can pass the same payload multiple times and get back
+    consistent results (idempotency, more or less).
+
+    To elaborate, as a use case consider a model like epi.models.Criteria;
+    this contains a unique_together meta restriction on the "assessment"
+    and "description" fields. Or in other words, in the database
+    assessment+description are guaranteed to be unique.
+
+    We want to build an API able to support something like a POST
+    { "assessment": 1, "description": "foo" }
+    to the endpoint defined for this. A normal serializer will complain
+    if a criteria with that description already exists, due to the
+    UniqueTogetherValidator firing. If instead we mix this class into
+    the appropriate serializer, we can code things such that the first call
+    will create a criteria with description "foo", and the second
+    call will just fetch the existing one. This makes client construction
+    just a little more straightforward -- rather than having to
+    lookup/check/create-if-needed, clients can just hit one endpoint and
+    get back the same object id every time.
+
+    Basic approach taken from https://stackoverflow.com/questions/25026034/django-rest-framework-modelserializer-get-or-create-functionality
+   """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        user = kwargs.get("context", {}).get("user")
```

Comment:
Since viewsets add `request` to the context by default you can access the user that way (`request.user`) instead of passing user in explicitly. That would simplify the `get_serializer` calls a little bit.